hivemind/client/averaging/allreduce.py

```diff
@@ -21,256 +22,203 @@ class AveragingMode(Enum):
     AUX = 2
 
 
-class AllReduceProtocol:
+class AllReduceProtocol(averaging_pb2_grpc.DecentralizedAveragingServicer):
     """
     An internal class that runs butterfly AllReduce in a predefined group of averagers
 
-    :param tensors: local tensors that should be averaged with groupmates
+    :note: this class returns **differences** between averaged and local tensors in order to improve numerical stability
```

Comment:
Alas, yes; it will complicate code in both partition.py and allreduce.py AND an additional lock that ensures that local tensors did not change.

Since #225, we already return deltas all the time and only leave an option to revert to non-deltas by adding them back. This behavior persists in the current PR.

Fortunately, 
- there are no drawbacks to sending deltas, the cost of subtraction is negligible compared to any(averaging, compression, transmission)
- if absolutely necessary for some reason, one can restore the average tensor by adding deltas to local tensors